mint-client/src/clients/user.rs

```diff
@@ -420,3 +452,190 @@ impl From<LnClientError> for ClientError {
+
+struct InvoiceVisitor {}
+struct PegOutReqVisitor {}
+
+impl<'de> serde::de::Visitor<'de> for InvoiceVisitor {
+    type Value = WrappedInvoice;
+
+    fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+        formatter.write_str("Could not deserialize Invoice")
+    }
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        match v.parse::<lightning_invoice::Invoice>() {
+            Ok(bolt11) => Ok(WrappedInvoice(bolt11)),
+            Err(e) => Err(serde::de::Error::custom(e.to_string())),
+        }
+    }
+}
+
+impl<'de> serde::de::Visitor<'de> for PegOutReqVisitor {
+    type Value = PegOutReq;
+
+    fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+        formatter.write_str("Could not deserialize pegout request")
+    }
+    fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: serde::de::MapAccess<'de>,
+    {
+        let mut addr: Option<&str> = None;
+        let mut amt: Option<u64> = None;
+
+        while let Some(key) = access.next_key::<String>()? {
+            match key.as_str() {
+                "address" => addr = Some(access.next_value()?),
+                "amount" => amt = Some(access.next_value()?),
+                _ => (),
+            }
+        }
+        Ok(PegOutReq {
+            address: bitcoin::Address::from_str(addr.unwrap()).unwrap(), //unwraps are not ok here I think but I will come back to Errors in another commit anyway
+            amount: bitcoin::Amount::from_sat(amt.unwrap()),
+        })
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for WrappedInvoice {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(InvoiceVisitor {})
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for PegOutReq {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_map(PegOutReqVisitor {})
+    }
+}
+///Holds either a Success or an Error variant and will be send as a response to the client (note: this means all [`ClientError`] will be send as [`tide::Result`] Ok
+/// only other errors will cause the server to respond with [`tide::Result`] Err)
+pub enum Event {
+    Success(APIResponse),
+    Error(APIResponse), //TODO: define Error APIResponse variant, implement a build method and use it in the endpoints
+}
+/// Holds all possible Responses of the RPC-CLient can also be used to parse responses (for client-cli)
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub enum APIResponse {
+    ///The clients holdings : The quantity of coins for each tier. For total holdings sum(Infoi.quantity * Infoi.tier) with i = 0 - n
+    /// Also contains the [`APIResponse::Pending`] variant.
+    Info {
+        coins: Vec<CoinsByTier>,
+        pending: Box<APIResponse>,
```

Comment:
Maybe better to also have a `Pending` struct instead of this cyclic reference where only one enum variant is valid? Accessing the pending value will be very annoying because you have to match the `ApiResponse` struct and everything that isn't the pending variant would panic.

Reply:
Mhm, I'll have to look into that (don't quite understand what you mean) I thought in this case I would have 2 APIResponses with one having a pointer in it to the other one (so not cyclic ?)

Reply:
The point I'm making is that a malformed `APIResponse` could have a `APIResponse::Info` variant in `pending` (or anything else that's unexpected). Am I right that this is only meant to contain `APIResponse::Pending`? If so there can be a struct `PendingResponse` that is the only data contained in `APIResponse::Pending(PendingResponse)` and also part of info.